middleware/grpc_cluster.go

```diff
@@ -6,48 +6,122 @@ import (
+	}
+	if invalidCluster == nil {
+		panic("no invalid cluster counter provided")
+	}
+	if logger == nil {
+		panic("no logger provided")
+	}
+}
+
+func handleError(err error, cluster string, method string, invalidCluster *prometheus.CounterVec, logger log.Logger) error {
+	if err == nil {
+		return nil
+	}
+	if stat, ok := grpcutil.ErrorToStatus(err); ok {
+		details := stat.Details()
+		if len(details) == 1 {
+			if errDetails, ok := details[0].(*grpcutil.ErrorDetails); ok {
+				if errDetails.GetCause() == grpcutil.WRONG_CLUSTER_VERIFICATION_LABEL {
+					msg := fmt.Sprintf("request rejected by the server: %s", stat.Message())
+					level.Warn(logger).Log("msg", msg, "method", method, "clusterVerificationLabel", cluster)
+					invalidCluster.WithLabelValues(method, clusterutil.ReasonServer).Inc()
+					return grpcutil.Status(codes.Internal, msg).Err()
+				}
+			}
+		}
 	}
+	return err
 }
 
-// ClusterUnaryServerInterceptor checks if the incoming gRPC metadata contains any cluster information and if so,
-// checks if the latter corresponds to the given cluster. If it is the case, the request is further propagated.
+// ClusterUnaryServerInterceptor checks if the incoming gRPC metadata contains any cluster label and if so, checks if
+// the latter corresponds to the given cluster label. If it is the case, the request is further propagated.
+// If an empty cluster label or nil logger are provided, ClusterUnaryServerInterceptor panics.
+// If the softValidation parameter is true, errors related to the cluster label validation are logged, but not returned.
 // Otherwise, an error is returned.
-func ClusterUnaryServerInterceptor(cluster string, logger log.Logger) grpc.UnaryServerInterceptor {
-	return func(ctx context.Context, req interface{}, _ *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
-		reqCluster := getClusterFromIncomingContext(ctx, logger)
-		if cluster != reqCluster {
-			msg := fmt.Sprintf("request intended for cluster %q - this is cluster %q", reqCluster, cluster)
-			level.Warn(logger).Log("msg", msg)
-			return nil, status.Error(codes.FailedPrecondition, msg)
+func ClusterUnaryServerInterceptor(cluster string, softValidation bool, logger log.Logger) grpc.UnaryServerInterceptor {
+	validateClusterServerInterceptorInputParameters(cluster, logger)
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		// We skip the gRPC health check.
+		if _, ok := info.Server.(healthpb.HealthServer); ok {
+			return handler(ctx, req)
 		}
-		return handler(ctx, req)
+
+		msgs, err := checkClusterFromIncomingContext(ctx, info.FullMethod, cluster, softValidation)
+		if len(msgs) > 0 {
+			level.Warn(logger).Log(msgs...)
+		}
+		if err == nil {
+			return handler(ctx, req)
+		}
+		stat := grpcutil.Status(codes.FailedPrecondition, err.Error(), &grpcutil.ErrorDetails{Cause: grpcutil.WRONG_CLUSTER_VERIFICATION_LABEL})
+		return nil, stat.Err()
```

Comment:
Nit just because that's the usual order.
```suggestion
		if err != nil {
			stat := grpcutil.Status(codes.FailedPrecondition, err.Error(), &grpcutil.ErrorDetails{Cause: grpcutil.WRONG_CLUSTER_VERIFICATION_LABEL})
			return nil, stat.Err()
		}
		return handler(ctx, req)
```